### PR TITLE
Ability to control the randomization

### PR DESCRIPTION
**Pull request by mebrunet**
These changes add some simple features to allow the user to control the randomization. Helpful for repeatability of experiments.

- add seed option to glove and shuffle
- add save/load initial parameter to glove

### CODE REVIEW

**Comment by AngledLuffa** [comment]
Weirdly the calculations are not exactly the same when running the same random seed on the same data.  I would have thought it would be, since even considering that floating point math is not exact, the same operations done in the same order should have the same result.

**Comment by mebrunet** [comment]
You need to limit the computation to a single thread for complete reproducibility. Because at the moment there is nothing managing the order in which to apply the updates from different threads. My understanding is that this is a common issue when trying to reproduce multi threaded/process computations. Worker management / ordering might be worth looking into, since the embedding process runs at a crawl with `-threads 1`. But at least with this PR reproducibility is possible. 

I'll see about whitespace.

**Comment by AngledLuffa** [comment]
Thank you... I missed that it was running multiple threads.  Of course that
makes sense.  I'll recheck with that in mind.

**Comment by AngledLuffa** [comment]
I checked in the random seed portion, as mentioned, and gave you credit.  Also added that seed = 0 means it chooses a seed based on the system time.

File: src/glove.c [hunk 1]
```diff
@@ -64,33 +67,46 @@ int scmp( char *s1, char *s2 ) {
...
+            if (feof(fin)) {
+                fprintf(stderr, "EOF reached before parameters fully loaded.\n");
+                exit(1);
+            }
+            fread(&W[a], sizeof(real), 1, fin);
         }
+        fclose(fin);
+    // Initialize new parameters
+    } else {
+        for (b = 0; b < vector_size; b++) for (a = 0; a < 2 * vocab_size; a++) W[a * vector_size + b] = (rand() / (real)RAND_MAX - 0.5) / vector_size;
```
Comment (AngledLuffa) [inline src/glove.c]: Just throwing this thought out there, bunching up lines like this makes it hard to read that diffs

File: src/glove.c [hunk 2]
```diff
@@ -229,7 +246,7 @@ int save_params(int nb_iter) {
         fid = fopen(vocab_file, "r");
         sprintf(format,"%%%ds",MAX_STRING_LENGTH);
         if (fid == NULL) {fprintf(stderr, "Unable to open file %s.\n",vocab_file); return 1;}
-        if (write_header) fprintf(fout, "%lld %d\n", vocab_size, vector_size);
+	      if (write_header) fprintf(fout, "%ld %d\n", vocab_size, vector_size);
```
Comment (AngledLuffa) [inline src/glove.c]: Any particular reason for this?

**Comment by AngledLuffa** [comment]
I think there should be some mechanism so that it doesn't save params if they were loaded

I'll add that

**Comment by AngledLuffa** [comment]
> I think there should be some mechanism so that it doesn't save params if they were loaded
> 
> I'll add that

or maybe not - it does make it easy to test and the parameters are different, at least

**Comment by AngledLuffa** [comment]
I love the fact that you included a test script!  I took the liberty of editing the test script to use $(mktemp ...) and of merging in other changes from other pull requests.  I made sure to give you credit when checking it in, though.  Thanks!

**Pull request closed without merging**